### PR TITLE
Err reform

### PR DESCRIPTION
**Pull request by graydon**
This is a WIP branch that's quite repetitive but ultimately fairly simple, that performs a variety of fixes related to events and errors. I _think_ it's all stuff we've already agreed "needs to happen" and I figured I'd just try to get it all done this week while everyone else was away from keyboard.

Hopefully I'll be done soon but it's going to probably take another day or two of just grinding through errors. Here's basically what it covers:

1. Renames `Status` to `Error` (the most-agreeable part of #689)
2. Reassigns all error codes to a smaller, consolidated set, fixing #225 (i.e. paired with change https://github.com/stellar/stellar-xdr/pull/92 where the set is explained, and a corresponding WIP rs-stellar-xdr change)
3. Deletes `DebugEvent` entirely, moving all diagnostics to structured-event form, answering at least _some_ of #731
4. As a result, assigns a causal contract ID to every debug event, fixing #772 (and #438 which is I think an earlier dupe of the same issue)
5. Also as a result, externalizes full objects when diagnostic mode is turned on, causing them to render fully, fixing #442
6. Fixes the double-faulting issue in #687
7. Adds a bunch more diagnostic context in cases where we're generating errors (needed since we're simplifying the codes), fixing #673

I thought I'd post it for feedback early, in case anyone _really_ hates the direction. I'm pretty keen on how it's turning out so far -- it seems like it really addresses a lot of the shortcomings we've identified so far and most of the churn is just obvious cleanup after the main structural changes -- but I'm interested to know if it looks to anyone like it's going the wrong direction.

### CODE REVIEW

File: soroban-env-host/src/auth.rs [hunk 1]
```diff
@@ -736,7 +758,12 @@ impl AuthorizationTracker {
                 Some(curr_invocation.sub_invocations.len() - 1);
         } else {
             // This would be a bug
-            return Err(host.err_general("unexpected missing authorized invocation"));
+            return Err(host.err(
+                ScErrorType::Auth,
+                ScErrorCode::MissingValue,
```
Comment (dmkozh) [inline soroban-env-host/src/auth.rs:763]: nit: this should be `InternalError` probably (given that this is an 'implementation bug' state).

The same actually applies to all the errors below besides 'nonce is incorrect' one. Basically anything starting with `unexpected` is an internal error and shouldn't ever happen (unless we have bugs).

File: soroban-env-host/src/events/mod.rs [hunk 2]
```diff
@@ -23,12 +31,110 @@ pub struct HostEvent {
...
+            u256_from_pieces(v.hi_hi, v.hi_lo, v.lo_hi, v.lo_lo)
+        ),
+        ScVal::I256(v) => write!(
+            f,
+            "{}",
+            i256_from_pieces(v.hi_hi, v.hi_lo, v.lo_hi, v.lo_lo)
+        ),
+        ScVal::Bytes(v) => write!(f, "Bytes({})", v.0),
+        ScVal::String(v) => write!(f, "\"{}\"", v.0),
+        ScVal::Symbol(v) => write!(f, "{}", v.0),
```
Comment (dmkozh) [inline soroban-env-host/src/events/mod.rs:72]: nit: Shall we do `Symbol({})` format for clarity? Othewise this seems a bit too easy to confuse with string

File: soroban-env-host/src/host/metered_map.rs [hunk 3]
```diff
@@ -123,8 +120,8 @@ where
                 ctx: Default::default(),
             })
         } else {
-            // TODO use a better error code for "unbounded input itertors"
-            Err(ScHostFnErrorCode::UnknownError.into())
+            // This is a logic error, we shoudl never get here.
```
Comment (dmkozh) [inline soroban-env-host/src/host/metered_map.rs:123]: s/shoudl/should

File: soroban-env-host/src/host/metered_vector.rs [hunk 4]
```diff
@@ -82,27 +87,22 @@ where
             vec.charge_deep_clone(budget)?;
             Self::from_vec(vec)
         } else {
-            // TODO use a better error code for "unbounded input iterators"
-            Err(ScHostFnErrorCode::UnknownError.into())
+            // This is a logic error, we shoudl never get here.
```
Comment (dmkozh) [inline soroban-env-host/src/host/metered_vector.rs:90]: s/shoudl/should

File: soroban-env-host/src/events/mod.rs (hunk 2 above)
Comment (graydon) [reply soroban-env-host/src/events/mod.rs:72]: I tried putting some kind of marker on it and it just reads badly. strings are quoted, so in practice it seems not too bad (and this is just for diagnostic display anyway, when the user gets an error dumped out on the console)

File: soroban-env-common/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+#[derive(Copy, Clone)]
+pub struct Error(RawVal);
+
+impl_wrapper_tag_based_rawvalconvertible!(Error);
+impl_wrapper_tag_based_constructors!(Error);
+impl_wrapper_as_and_to_rawval!(Error);
+impl_wrapper_wasmi_conversions!(Error);
+
+impl Error {
+    pub const UNKNOWN: Error =
```
[the remark this answers is not in the document]
Comment (graydon) [reply soroban-env-common/src/error.rs:29]: Between unknown and internal? Fair point. I'll see how it looks if I remove unknown entirely.

File: soroban-env-host/src/auth.rs (hunk 1 above)
Comment (graydon) [reply soroban-env-host/src/auth.rs:763]: ok I'll fix these

File: soroban-env-common/src/error.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (graydon) [reply soroban-env-common/src/error.rs:29]: Yeah, this works fine with just "internal", changing.

File: soroban-env-host/src/auth.rs (hunk 1 above)
Comment (graydon) [reply soroban-env-host/src/auth.rs:763]: fixed (all except the ones about nonces in the ledger having the wrong type -- that seems like a user error)

File: soroban-env-host/src/host/metered_vector.rs (hunk 4 above)
Comment (graydon) [reply soroban-env-host/src/host/metered_vector.rs:90]: fixed

File: soroban-env-host/src/host/metered_map.rs (hunk 3 above)
Comment (graydon) [reply soroban-env-host/src/host/metered_map.rs:123]: fixed

File: soroban-env-host/src/auth.rs (hunk 1 above)
Comment (dmkozh) [reply soroban-env-host/src/auth.rs]: These would actually be internal as well as the host manages nonces exclusively currently. But nonces also will hopefully change the implementation after temp storage introduction, so also feel free to skip these for now, it doesn't really matter that much.

File: soroban-env-common/src/error.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-common/src/error.rs]: Nice, thanks!

File: soroban-env-common/src/env.rs [hunk 6]
```diff
@@ -137,65 +137,10 @@ pub trait EnvBase: Sized + Clone {
...
-    fn log_static_fmt_general(
-        &self,
-        fmt: &'static str,
-        vals: &[RawVal],
-        strs: &[&'static str],
-    ) -> Result<(), Self::Error>;
+    /// Log a string and set of values as a diagnostic event, if diagnostic
+    /// events are enabled. When running on host, logs directly; when running on
+    /// guest, redirects through log_from_linear_memory.
+    fn log_from_slice(&self, msg: &str, vals: &[RawVal]) -> Result<Void, Self::Error>;
```
[the remark this answers is not in the document]
Comment (graydon) [reply soroban-env-common/src/env.rs:143]: Yeah, the thing we were doing before where guest.rs turned the log calls into no-ops when compiling for wasm we're no longer doing. I removed basically all the special static cases around diagnostic logging -- even static-string optimization in events -- and just went with a 100% dynamic-strings and dynamic-vecs solution. It's costly memory-wise but you don't get to incur those costs in any context where it'll do harm: we'll prohibit turning on dynamic logging on validators entirely, and set it default-off on all production nodes. Users can turn it on on production nodes tx-by-tx if they want, but not validators. And any local case, obviously -- in the SDK change I figured you'd just turn it on by default, but we can discuss that when reviewing the SDK change (I expect you'll have opinions there!)

**Pull request merged**